e4s: add power stack

eugeneswalker opened the pull request:
In collaboration with @scottwittenburg -- we have no power nodes in the protected infrastructure set -- going to try and find a way to make E4S Power builds happen on PRs while skipping the post-merge build and cache step that requires protected runner infrastructure.

FYI @wspear

eugeneswalker committed:
e4s: restore builds builds

eugeneswalker commented:
GitLab pipeline all green now! Just need to figure out how we should handle the protected builds. Eager to get Power builds back in the loop-- ensuring they stay buildable as new PRs come in to Spack.

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml:
```diff
@@ -0,0 +1,424 @@
+          variables:
+            CI_JOB_SIZE: huge
+            KUBERNETES_CPU_REQUEST: 11000m
+            KUBERNETES_MEMORY_REQUEST: 42G
```
I'm of two minds on including these variables in the mappings.  They are used by kubernetes in scheduling job onto pods running on aws runners.  Since we don't have any power runners in the aws cluster (and might not ever), these variables are unused in this stack and might just confuse people.  However, I could see keeping them here (and making them somewhat accurate) for consistency with other stacks and/or in case we ever got power runners in the cloud.  

I'm not requesting a change here, just thinking out loud.

eugeneswalker replied:
Absolutely! Good catch. I didn't mean to include these -- I was just copying the other environment and then modifying. I will remove them since they are not needed. Totally in favor of keeping things as simple as possible. Thank you.

scottwittenburg commented:
@zackgalbreath @kwryankrattiger Added you as reviewers to get more eyes on the small refactor I made to `.gitlab-ci.yml`.

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
Only failure blocking us now is `aws-isc-aarch64-pr-generate` pipeline, which is failing due to broken spec on develop.

eugeneswalker commented:
> Only failure blocking us now is `aws-isc-aarch64-pr-generate` pipeline, which is failing due to broken spec on develop.

Just rebased on develop, hoping this time nothing is broken on develop.

eugeneswalker commented:
> > Only failure blocking us now is `aws-isc-aarch64-pr-generate` pipeline, which is failing due to broken spec on develop.
> 
> Just rebased on develop, hoping this time nothing is broken on develop.

Rebased again on develop. I submitted a separate PR earlier today, and no specs were reported as broken for `aws-isc-aarch64-pr-generate`

eugeneswalker commented:
Broken spec on develop being detected by `radiuss-aws-pr-generate` now, others have resolved....

eugeneswalker commented:
@spackbot run pipeline because system failures are causing jobs to fail erroneously

eugeneswalker commented:
All checks are GREEN!

What do you think @scottwittenburg -- ready to merge?

scottwittenburg commented:
Do you know what happened to the refactoring of .`gitlab-ci.yml` I pushed on your branch 9 days ago?

scottwittenburg committed:
gitlab ci: allow UO to build protected binaries for signing

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

scottwittenburg commented:
fyi @kwryankrattiger

eugeneswalker committed:
use newer image; comment out failing builds

eugeneswalker commented:
@scottwittenburg How should we proceed from here? I've lost track of the state of this. It looks like none of the CI checks have run since my last commit ~1 week ago...

scottwittenburg commented:
I'm guessing the ci checks don't run if there's a merge conflict?  Want me to look into the conflicts?  I'm not sure how the power stack changed recently when it's been disabled forever, but I can dig into it if you want.

eugeneswalker commented:
> I'm guessing the ci checks don't run if there's a merge conflict? Want me to look into the conflicts? I'm not sure how the power stack changed recently when it's been disabled forever, but I can dig into it if you want.

I think I just fixed the conflicts. If all CI checks pass, do you think this will be good to go? I would assume yes, except that I know there've been some big changes in CI at least behind the scenes with the merge queue and whatnot, wondering if any of that affects what we're trying to do here. I don't know. Thanks!

scottwittenburg commented:
Looking over it again just now, I *think* it's good to go.  However, since we can't actually do a trial run on the protected jobs, that's the only part I'm concerned about.  A couple options there:

1. Add a failure allowance to those jobs now and plan to remove it once we see everything working as expected
2. Keep it the way it is now, and plan to add a temporary failure allowance if those jobs fail right away

If we go with option (1), then we can merge this and check back as time permits.  If we go with (2), we may want to keep a close eye on the protected pipelines after merging this to be ready to act.  I'm ok with either approach.

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
I just updated the E4S power environment to include only specs that are known to build OK in the container image:
* `ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01` 

I just did a full rebuild from source, in this container, and everything in this environment built OK. I added notes for the failing specs.

I also just connected a number of Power9 runners to gitlab.spack.io which I'm hoping can handle the builds for this PR. These runners have protected runner equivalents at UO which I've paused temporarily, only planning to turn those on if this looks like it is ready to pass the PR stage and get merged. Just FYI. This has been a long time coming and I think thanks to help from the kitware team we are finally very close.

eugeneswalker commented:
CI only failing because of an issue unrelated to this PR:
* [aws-ahug-aarch64-pr-generate](https://gitlab.spack.io/spack/spack/-/jobs/5258648)

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
> CI only failing because of an issue unrelated to this PR:
> 
>     * [aws-ahug-aarch64-pr-generate](https://gitlab.spack.io/spack/spack/-/jobs/5258648)

Latest pipeline run failing for same reason.

scottwittenburg committed:
gitlab-ci: Some tweaks for e4s power builds

- fix tags (no longer require generate jobs to run on aws)
- fix resource requests for generation jobs resource requests
- remove SPACK_SIGNING_KEY from protected power build jobs
- update UO signing key path
- change the CDash build group to reflect stack name
- retry pipeline generation jobs *always*

scottwittenburg commented:
@kwryankrattiger This may get merged before #34272 and if so, could you rebase on `develop` and incorporate this stack into your changes there?  Pretty please?  :grin: 

On the other hand, @eugeneswalker, if his PR gets merged first, we'll need to update this to remove all the boilerplate and make it look like the other stacks post-34272.  I'm happy to do it, if that's how things work out.

kwryankrattiger commented:
Either way I will translate all of this in my PR, that way it should just be there but not activated yet in the root `.gitlab-ci.yaml` file.

This has been a really long running update so it will be exciting to see it finally make it in!

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml:
```diff
@@ -0,0 +1,460 @@
+      - mkdir -p ${SPACK_ARTIFACTS_ROOT}/user_data
+      - if [[ -r /mnt/key/e4s.gpg ]]; then spack gpg trust /mnt/key/e4s.gpg; fi
+      - if [[ -r /mnt/key/spack_public_key.gpg ]]; then spack gpg trust /mnt/key/spack_public_key.gpg; fi
+      - spack -d ci rebuild > >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt) 2> >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt >&2)
```
Is it just copy-pasta or is there a specific reason this script is different.

Missing/different components are:
* downloading make for parallel builds
* `spack -d` instead of `spack --color=always --backtrace`
* no `--test`
* skips printing system info (`uname`, `cpuinfo`, `nproc`)
* excludes checking for the `/mnt/key/intermediate_ci_signing_key.gpg` key

eugeneswalker replied:
Good catch. I just pushed a change to include all the aspects you mention except the intermediate ci signing key, which is not needed. The reason it isn't needed for this stack is because only UO hardware is going to run these power builds. There are no power runners on AWS. @scottwittenburg or @kotfic could explain in more depth but that's my understanding.

EDIT: also, we are not going to run `--tests` for the power stack out of the gate. We may pursue that in the future but not in the first pass.

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml:
```diff
@@ -0,0 +1,460 @@
+
+    image: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
+
+    match_behavior: first
```
These mappings seem to be updated from the old power stack, I am assuming this is a copy-pasta and replace  `x86_64` -> `ppc64le`?

eugeneswalker replied:
I wouldn't describe it that way. Fundamentally, it's just simply that because these builds are happening on power machines, we need to use an image that works for power machines :)

eugeneswalker committed:
correct double packages: section

eugeneswalker committed:
gitlab-ci:script: modernize

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml:
```diff
@@ -0,0 +1,426 @@
+      - uname -a || true
+      - grep -E 'vendor|model name' /proc/cpuinfo 2>/dev/null | sort -u || head -n10 /proc/cpuinfo 2>/dev/null || true
+      - nproc
+      - curl -Lfs 'https://github.com/JuliaBinaryWrappers/GNUMake_jll.jl/releases/download/GNUMake-v4.3.0+1/GNUMake.v4.3.0.x86_64-linux-gnu.tar.gz' -o gmake.tar.gz
```
I think maybe this is for x86_64 only. Either `ci rebuild` needs an option for the make stuff or this may need to get make in another way.

eugeneswalker replied:
yes, you are right. will fix.

eugeneswalker committed:
remove new gnu make, not for ppc64le

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
All CI is finally green! Can someone give the final manual approval so we can get this merged before anything changes? :)

kwryankrattiger reviewed (commented):
LGTM, thanks @eugeneswalker and @scottwittenburg for all of your work on this!
